homeassistant/helpers/config_validation.py

```diff
@@ -518,19 +539,127 @@ def ensure_list_csv(value: Any) -> Sequence:
     return ensure_list(value)
 
 
-def deprecated(key):
-    """Log key as deprecated."""
+def deprecated(key: str,
+               replacement_key: Optional[str] = None,
+               invalidation_version: Optional[str] = None,
+               default: Optional[Any] = None):
+    """
+    Log key as deprecated and provide a replacement (if exists).
+
+    Expected behavior:
+        - Outputs the appropriate deprecation warning if key is detected
+        - Processes schema moving the value from key to replacement_key
+        - Processes schema changing nothing if only replacement_key provided
+        - No warning if only replacement_key provided
+        - No warning if neither key nor replacement_key are provided
+            - Adds replacement_key with default value in this case
+        - Once the invalidation_version is crossed, raises vol.Invalid if key
+        is detected
+    """
     module_name = inspect.getmodule(inspect.stack()[1][0]).__name__
 
-    def validator(config):
+    if replacement_key and invalidation_version:
+        warning = ("The '{key}' option (with value '{value}') is deprecated,"
+                   " please replace it with '{replacement_key}'. This option"
+                   " will become invalid in version {invalidation_version}.")
+    elif replacement_key:
+        warning = ("The '{key}' option (with value '{value}') is deprecated,"
+                   " please replace it with '{replacement_key}'.")
+    elif invalidation_version:
+        warning = ("The '{key}' option (with value '{value}') is deprecated,"
+                   " please remove it from your configuration. This option"
+                   " will become invalid in version {invalidation_version}.")
+    else:
+        warning = ("The '{key}' option (with value '{value}') is deprecated,"
+                   " please remove it from your configuration.")
+
+    def check_for_invalid_version(value: Optional[str]):
```

Comment:
At some point, it may be worth pulling this out of an inner function and making it a decorator that could be used in other things besides configuration expiration.